ws/api/endpoints.go

```diff
@@ -0,0 +1,186 @@
+	"google.golang.org/grpc/status"
+)
+
+var channelPartRegExp = regexp.MustCompile(`^/channels/([\w\-]+)/messages(/[^?]*)?(\?.*)?$`)
+
+func handshake(svc ws.Service) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		req, err := decodeRequest(r)
+		if err != nil {
+			encodeError(&req, w, err)
+			return
+		}
+		conn, err := upgrader.Upgrade(w, r, nil)
+		if err != nil {
+			logger.Warn(fmt.Sprintf("Failed to upgrade connection to websocket: %s", err.Error()))
+			return
+		}
+		req.conn = conn
+		client := ws.NewClient(conn)
+
+		if err := svc.Subscribe(context.Background(), req.thingKey, req.chanID, req.subtopic, client); err != nil {
+			logger.Warn(fmt.Sprintf("Failed to subscribe to broker: %s", err.Error()))
+			req.conn.Close()
+			return
+		}
+
+		logger.Debug(fmt.Sprintf("Successfully upgraded communication to WS on channel %s", req.chanID))
+		msgs := make(chan []byte)
+
+		// Listen for messages received from the chan messages, and publish them to broker
+		go process(svc, &req, msgs)
+		go listen(conn, msgs)
+	}
+}
+
+func decodeRequest(r *http.Request) (connReq, error) {
+	authKey := r.Header.Get("Authorization")
+	if authKey == "" {
+		authKeys := bone.GetQuery(r, "authorization")
+		if len(authKeys) == 0 {
+			logger.Debug("Missing authorization key.")
+			return connReq{}, errUnauthorizedAccess
+		}
+		authKey = authKeys[0]
+	}
+
+	chanID := bone.GetValue(r, "id")
+
+	req := connReq{
+		thingKey: authKey,
+		chanID:   chanID,
+	}
+
+	channelParts := channelPartRegExp.FindStringSubmatch(r.RequestURI)
+	if len(channelParts) < 2 {
+		logger.Warn("Empty channel id or malformed url")
+		return connReq{}, errors.ErrMalformedEntity
+	}
+
+	subtopic, err := parseSubTopic(channelParts[2])
+	if err != nil {
+		return connReq{}, err
+	}
+
+	req.subtopic = subtopic
+
+	return req, nil
+}
+
+func parseSubTopic(subtopic string) (string, error) {
+	if subtopic == "" {
+		return subtopic, nil
+	}
+
+	subtopic, err := url.QueryUnescape(subtopic)
+	if err != nil {
+		return "", errMalformedSubtopic
+	}
+
+	subtopic = strings.Replace(subtopic, "/", ".", -1)
+
+	elems := strings.Split(subtopic, ".")
+	filteredElems := []string{}
+	for _, elem := range elems {
+		if elem == "" {
+			continue
+		}
+
+		if len(elem) > 1 && (strings.Contains(elem, "*") || strings.Contains(elem, ">")) {
+			return "", errMalformedSubtopic
+		}
+
+		filteredElems = append(filteredElems, elem)
+	}
+
+	subtopic = strings.Join(filteredElems, ".")
+
+	return subtopic, nil
+}
+
+func listen(conn *websocket.Conn, msgs chan<- []byte) {
+	for {
+		// Listen for message from the client, and push them to the msgs channel
+		_, payload, err := conn.ReadMessage()
+
+		if websocket.IsUnexpectedCloseError(err) {
+			logger.Debug(fmt.Sprintf("Closing WS connection: %s", err.Error()))
+			close(msgs)
+			return
+		}
+
+		if err != nil {
+			logger.Warn(fmt.Sprintf("Failed to read message: %s", err.Error()))
+			close(msgs)
+			return
+		}
```

Comment:
These scenarios are not tested, which also results in L153-L156 never executing during tests. It's important to test API layer extensively because the entire protocol support happens here.

Reply:
These are tested manually, when I force close or suspend and close, either the terminal, or the ws-client.
And then connect to them again, and in both cases the closure is handled by the ws-adapter.
I'm able to normally connect back using the previously unexpectely closed client, and continue to send and receive messages.